### PR TITLE
feat(fuzz): add XSS reflection context analyzer

### PR DESCRIPTION
**Pull request by CharlesWong**
## Summary

/claim #5838

Adds `pkg/fuzz/analyzers/xss/` — a standalone XSS context analyzer that determines **where** in an HTML document a reflected fuzz marker lands, then returns context-appropriate exploit payloads with confidence scores.

Fixes the `javascript:` URI misclassification described in #7086.

---

## Problem

The fuzzing engine finds reflections but has no way to tell *where* in the HTML the value lands. Without context, you either spray every XSS payload everywhere (slow, noisy, false-positive-heavy) or miss valid injection points entirely.

---

## Solution

`AnalyzeReflectionContext(responseBody, marker string) XSSResult`

Uses `golang.org/x/net/html` tokenizer (already in `go.mod`) to walk the HTML token stream. Returns a full `XSSResult` with:

- **Context** (16 types — see table below)
- **Payloads** — ordered list of exploit candidates best suited to the context
- **BreakoutSeq** — minimal sequence to escape the current context
- **Confidence** — 0.0–1.0 score
- **IsExecutableSink** — true when JS runs directly without additional breakout
- **Explanation** — human-readable description
- **QuoteChar**, **AttributeName**, **TagName** — structural metadata

`AnalyzeAllReflections` scans the entire document for all occurrences; `AnalyzeReflectionContext` returns the most exploitable one.

Implements the `analyzers.Analyzer` interface (registered as `"xss_context"`) so it can be referenced in template YAML.

---

## Context types (16)

| Context | Example |
|---|---|
| `HTMLBody` | `<p>MARKER</p>` |
| `Comment` | `<!-- MARKER -->` |
| `AttributeDouble` | `<tag attr="MARKER">` |
| `AttributeSingle` | `<tag attr='MARKER'>` |
| `AttributeUnquoted` | `<tag attr=MARKER>` |
| `AttributeURL` | `<a href="MARKER">` |
| `AttributeEvent` | `<div onclick="MARKER">` |
| `AttributeStyle` | `<div style="MARKER">` |
| `Script` | `<script>MARKER</script>` or executable `javascript:` URI |
| `ScriptData` | `<script type="application/json">MARKER` |
| `Style` | `<style>MARKER { }\</style>` |
| `JSON` | `{"key": "MARKER"}` inside a script block |
| `Template` | `\`hello ${MARKER}\`` in a script block |
| `CDATA` | `<![CDATA[MARKER]]>` |
| `SrcDoc` | `<iframe srcdoc="MARKER">` |
| `Unknown` | marker not found or context unparseable |

---

## Key correctness decisions

**javascript: URI executability is tag-aware** (fixes #7086):
- `<a href="javascript:">` → `ContextScript` ✓ (executes)
- `<iframe src="javascript:">` → `ContextScript` ✓ (executes)
- `` → `ContextAttributeURL` ✓ (does NOT execute)
- `<script src="javascript:">` → `ContextAttributeURL` ✓ (does NOT execute)

**First `type` attribute wins on `<script>`** (HTML5 spec):
- `<script type="application/json" type="text/javascript">` → `ContextScriptData` (non-executable)

**TagAttr() single-pass** — the tokenizer's attribute iterator is forward-only. Scanning in one pass prevents silent attribute dropping (e.g. `<script src="MARKER">` would be missed in a two-pass approach).

**Quote style detection** — the raw HTML source is probed to determine the actual delimiter (`"`, `'`, or unquoted) so breakout sequences are correct.

---

## Test coverage (55 test cases + 3 benchmarks, all pass)

```
$ go test ./pkg/fuzz/analyzers/xss/... -v
...
PASS
ok  github.com/projectdiscovery/nuclei/v3/pkg/fuzz/analyzers/xss
```

Covers:
- All 16 context types
- javascript: URI across 7 different tag+attr combos
- Event handler with single/double quote variants
- First-type-wins for duplicate `type` attrs on `<script>`
- Case-insensitive marker matching
- Nested quotes (double inside single and vice versa)
- Template literal detection
- Multiple reflections in one document (best context selected)
- Empty / missing inputs

**Benchmarks (Apple M4):**
```
BenchmarkAnalyzeReflectionContext_HTMLBody        898772   1267 ns/op    4896 B/op    21 allocs/op
BenchmarkAnalyzeReflectionContext_LargeDocument     9140  131626 ns/op  44468 B/op  1815 allocs/op
BenchmarkAnalyzeAllReflections                    677701   2145 ns/op    6208 B/op    32 allocs/op
```

---

## Checklist
- [x] PR created against `dev` branch
- [x] New package only — no modifications to existing files
- [x] Tests added (55 cases + 3 benchmarks)
- [x] All tests pass
- [x] `go build ./pkg/fuzz/...` succeeds
- [x] Implements `analyzers.Analyzer` interface
- [x] `javascript:` URI misclassification (#7086) fixed with tag-aware sink map

## Summary by CodeRabbit

* **New Features**
  * New XSS reflection analyzer: detects and ranks reflections across HTML and JSON contexts (body, comments, attributes, scripts, styles, templates, srcdoc), reports confidence, and suggests context-aware proof‑of‑concept payloads and breakout guidance when exploitable reflections are found. Integrated into the scanning flow for automated analysis.

* **Tests**
  * Comprehensive test suite covering context detection, prioritization, payloads, edge cases, and benchmarks.

### CODE REVIEW

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,783 @@
...
+
+		// HTML5 spec: first `type` attribute wins when duplicates exist.
+		if attrName == "type" && !foundFirstType {
+			scriptType = strings.ToLower(strings.TrimSpace(attrValue))
+			foundFirstType = true
+		}
+
+		// Check if marker is in the attribute value
+		if containsMarkerStr(attrValue, markerLower) {
+			quoteChar := detectAttrQuoteChar(rawBody, markerOriginal, attrName)
```
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/analyzer.go:410]: Good catch. The quote detection should be scoped per-reflection rather than scanning the full body. Will fix to search backward from the reflection position to find the containing attribute's quote style.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,783 @@
...
+	// Additional check: must see a colon (JSON key separator) in the context
+	// Avoids matching JS: var x = "MARKER"
+	if last == '"' || last == '\'' {
+		// Look for a colon before the quote that opened the value
+		// We search backwards for the opening quote and check there's a : before it
+		// Simplified: check for `: "` or `: '` pattern
+		if !strings.Contains(trimmed, ": \"") && !strings.Contains(trimmed, ": '") &&
+			!strings.Contains(trimmed, ":{\"") && !strings.Contains(trimmed, ":\"") {
+			return false
+		}
```
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/analyzer.go:584]: Agreed — JSON uses only double quotes. Will add a pre-check: if the surrounding context uses single quotes, it cannot be ContextJSON and should be reclassified as ContextScript or ContextAttribute.

File: pkg/fuzz/analyzers/xss/xss_analyzer.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+func (a *XSSAnalyzer) Name() string {
+	return "xss_context"
+}
+
+// ApplyInitialTransformation returns the payload unchanged.
+// XSS context analysis does not pre-transform payloads; it classifies
+// reflections after the fact.
+func (a *XSSAnalyzer) ApplyInitialTransformation(data string, params map[string]interface{}) string {
+	return analyzers.ApplyPayloadTransformations(data)
+}
```
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/xss_analyzer.go:34]: Valid point. Will change to prefer `gr.Value` (the transformed fuzz value actually sent) over `gr.OriginalPayload` so the canary search matches what the server actually received.

File: pkg/fuzz/analyzers/xss/xss_analyzer.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		return false, "", nil
+	}
+
+	// Quick pre-check before full HTML parse
+	if !strings.Contains(strings.ToLower(string(bodyBytes)), strings.ToLower(marker)) {
+		return false, "", nil
+	}
+
+	result := AnalyzeReflectionContext(string(bodyBytes), marker)
+	if result.Context == ContextUnknown {
```
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/xss_analyzer.go:116]: Good suggestion. Will add Content-Type check: only run HTML context analysis when response Content-Type contains `text/html`. For JSON responses, skip directly to ContextJSON detection.

**Comment by CharlesWong** [comment]
Fixed all 4 coderabbitai issues in latest commit (2ba1f3c):

- **Quote detection now per-reflection**: `detectAttrQuoteChar` now searches backward from the reflection position using `strings.LastIndex` on the segment before the marker, instead of a global `strings.Index` on the whole body — eliminating false positives from same-name attributes elsewhere in the document.
- **JSON classification guarded against single-quote context**: Added an early-return in `isJSONContext` — if the character immediately before the marker is a single quote, it cannot be valid JSON (JSON only uses double quotes), so we return `false` and let the context fall through to `ContextScript` or `ContextAttribute`.
- **Canary search uses `gr.Value` (transformed payload)**: Changed `xss_analyzer.go` to use `gr.Value` as the primary marker (the actual transformed value sent to the server), falling back to `gr.OriginalPayload` only when `gr.Value` is empty.
- **HTML context analysis gated on `Content-Type: text/html`**: For non-HTML responses (e.g. `application/json`), the full HTML tokenizer is skipped; only JSON context classification is attempted, avoiding false context detection in structured data responses.

File: pkg/fuzz/analyzers/xss/analyzer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/analyzer.go:410]: Fixed in 2ba1f3c. `detectAttrQuoteChar` now walks backward from the specific marker occurrence using `LastIndex`, scoping quote detection per-reflection.

File: pkg/fuzz/analyzers/xss/analyzer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/analyzer.go:584]: Fixed in 2ba1f3c. `isJSONContext` now rejects single-quote contexts (`last != '"' && last != '['` → return false). Single-quoted values fall through to ContextScript/ContextAttribute.

File: pkg/fuzz/analyzers/xss/xss_analyzer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/xss_analyzer.go:34]: Fixed in 2ba1f3c. `Analyze()` now uses `gr.Value` first, falling back to `gr.OriginalPayload` only when Value is empty.

File: pkg/fuzz/analyzers/xss/xss_analyzer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/xss_analyzer.go:116]: Fixed in 2ba1f3c. `Analyze()` checks Content-Type before HTML tokenization; non-HTML responses skip the HTML path and only check JSON context.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+				if inScript {
+					ctx := ContextScriptData
+					if scriptIsExec {
+						ctx = ContextScript
+						// Refine: is it inside a template literal?
+						if isInTemplateLiteral(data, markerLower) {
+							ctx = ContextTemplate
+						} else if isJSONContext(data, markerLower) {
+							ctx = ContextJSON
+						}
```
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/analyzer.go]: Fixed in c9bb06f. The TextToken handler now loops with `strings.Index` from an advancing offset so every occurrence of the marker in a single token gets its own per-reflection analysis. For each match we pass the segment up to that occurrence to `isInTemplateLiteral` and `isJSONContext`, so each reflection independently resolves to the correct context — e.g. the second reflection in `var a="FUZZ"; var b=\`FUZZ\`` is no longer dropped.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+				`" onmouseover="alert(1)`,
+				`"><img src=x onerror=alert(1)>`,
+			)
+		} else if f.quoteChar == '\'' {
+			r.BreakoutSeq = `'`
+			r.Payloads = append(r.Payloads,
+				`' onmouseover='alert(1)`,
+			)
+		}
+		r.Explanation = "Marker is in a URL attribute. Try javascript: scheme or break out of the attribute."
```
[the remark this answers is not in the document]
Comment (CharlesWong) [reply pkg/fuzz/analyzers/xss/analyzer.go]: Fixed in c9bb06f. The `ContextAttributeURL` case in `buildResult` now branches on `f.isExecutable`: executable sinks (a[href], area[href], form[action], etc. — identified via the existing `executableURLSinks` map in `classifyAttrValueContext`) get `javascript:` payloads and `IsExecutableSink=true`; non-executable sinks like `img[src]` and `script[src]` get open-redirect payloads only (`//attacker.com/`) with no `javascript:` URIs surfaced.

**Review by CharlesWong — COMMENTED** [review]
Thanks for the nitpick review! Addressing all 4 items:

1. **L363 `_ = currentTag` no-op** — will remove, it was left as a placeholder for future tag-aware text-node payload selection that isn't implemented yet
2. **L819-824 dead code block** — will remove the empty if/else-if branches  
3. **L580-588 dead single-quote patterns in JSON check** — will simplify to `if last == '"'` only, consistent with RFC 8259 (JSON only uses double quotes)
4. **ContentType: missing XHTML + body sniff** — the 18:36 comment was marked duplicate (already-resolved), but the newer inline check at L81-83 does need the XHTML and sniff fix

Pushing a cleanup commit shortly.

**Review by CharlesWong — COMMENTED** [review]
All 4 nitpicks addressed in commit `64c5662`:

1. **`currentTag` unused variable** — removed the variable declaration entirely (was set at StartTagToken/EndTagToken but never read; the `_ = currentTag` no-op was the only "use")
2. **Dead code block** — removed the empty `if/else-if` for quote context refinement (lines 819-824)
3. **Single-quote patterns in JSON check** — simplified to `if last == '"'` only; the preceding `last == '\'' → return false` guard already makes single-quote branches unreachable, and RFC 8259 specifies JSON strings use only double quotes
4. **Content-Type: XHTML + body sniff** — added `application/xhtml+xml` to the HTML check and added a 512-byte HTML sniff for empty Content-Type (checking for `<!doctype`, `<html`, `<head`, `<body`, `<script`) instead of blindly assuming HTML

**Comment by songfrank13141-dotcom** [comment]
## Review Comments Addressed

I've addressed the remaining nitpick review comments from CodeRabbit. The following fixes have been implemented and pushed to my fork (available for cherry-picking):

### Changes Made

1. **Updated doc comment for `AnalyzeReflectionContext`** (analyzer.go:203-211)
   - Changed documentation to accurately reflect that the function returns the **most exploitable** context (best context wins) rather than the first match
   - References `AnalyzeAllReflections` and `contextPriority` as requested

2. **Removed redundant `inStyle` assignment** (analyzer.go:329-331)
   - Deleted duplicate `if tt == html.StartTagToken && tagName == "style" { inStyle = true }` block
   - Style state is now only set in the switch statement alongside script state handling

3. **Simplified `isJSONContext` single-quote rejection** (analyzer.go:563-573)
   - Changed initial guard from `if last != '"' && last != '\'' && last != '['` to `if last != '"' && last != '['`
   - Removed redundant subsequent single-quote rejection block
   - Now rejects single quotes immediately per RFC 8259 (JSON only uses double quotes)

4. **Fixed `detectAttrQuoteChar` for multiple marker occurrences** (analyzer.go:486-526)
   - Updated to iterate through all marker occurrences instead of anchoring to the first one
   - For each occurrence, checks if `attrName=` appears before it
   - Uses the first occurrence where the attribute matches, ensuring correct quote detection when the same marker appears in multiple attributes with different quote styles (e.g., `<a href='/x'></a><a href="MARKER">`)

### Commit Reference

These changes are available at: https://github.com/songfrank13141-dotcom/nuclei/commit/efb5525a

**Commit message:** `fix: address CodeRabbit nitpick reviews`

---

@CharlesWong Feel free to cherry-pick these changes or let me know if you'd like me to create a PR to your branch directly (if maintainer edits are enabled).

**Comment by Mzack9999** [comment]
Closing this PR as Issue #5838 has been resolved via PR #7164. While this is a well-structured implementation with good test coverage, the merged solution provides the required XSS context analysis with a more focused scope and proper integration with the existing analyzer framework. Thank you for the thorough work.
